extractor/extractor.py

```diff
@@ -0,0 +1,223 @@
+            text.text_string = text_data
+            return 1, text
+    elif type(node) == bs4.element.NavigableString and not empty_text(node) and len(node) > 50:
+        node = node.strip()
+        text = contents.Text()
+        text.type = ''
+        text.text_string = node
+        return 1, text
+
+    return 0, {}
+
+
+def get_image_size(url):
+    global req_session
+    try:
+        img = PIL.Image.open(req_session.get(url).text)
+        return img.size
+    except ValueError:
+        pass
+
+    return 0, 0
+
+
+def get_image_attributes(node,image):
+    image.img_url = node['src']
+    if node.has_attr('title'):
+        image.img_title = node['title']
+    h, w = get_image_size(image.img_url)
+    image.img_height = h
+    image.img_width = w
+    if image.img_url.endswith('.gif'):
+        image.is_gif = True
+
+
+def is_image(node):
+    if node.name == 'img' and node.has_attr('src'):
+        image = contents.Image()
+        get_image_attributes(node, image)
+        return 1, image
+
+    if node.name == 'figure':
+        img_tag = node.find('img')
+        if img_tag is not None and img_tag.has_attr('src'):
+            caption = node.find('figcaption')
+            image = contents.Image()
+            get_image_attributes(img_tag, image)
+            image.img_caption = caption
+            return 1, image
+
+    return 0, {}
+
+
+def get_video_attributes(node, video):
+    if node.has_attr('height') and node.has_attr('width'):
+        video.height = node['height']
+        video.width = node['width']
+    if node.has_attr('src'):
+        video.video_urls.append(node['src'])
+    elif len(node.contents) > 0:
+        sources = node.contents
+        for source in sources:
+            if source.name == 'source':
+                video.video_urls.append(source['src'])
+
+
+def is_video(node):
+    if node.name == 'video' and (node.has_attr('src') or len(node.contents) > 0):
+        video = contents.Video()
+        get_video_attributes(node, video)
+        return 1, video
+    elif node.name == 'embed' and node.has_attr('src'):
+        src = node['src']
+        pattern_for_video_ext = re.compile(r'^.*\.(mp4|ogg|mov|webm)')
+        is_match = pattern_for_video_ext.match(src)
+        if is_match is not None:
+            video = contents.Video()
+            if node.has_attr('height') and node.has_attr('width'):
+                video.height = node['height']
+                video.width = node['width']
+            video.video_urls.append(src)
+            return 1, video
+
+    return 0, {}
+
+
+def clean_soup(soup):
+    decomposable_tags = soup.find_all(['script', 'noscript', 'style', 'header', 'footer', 'meta', 'link'])
+    comments = soup.find_all(text=lambda text:isinstance(text, Comment))
+    ads = soup.find_all(re.compile('.*'), {'class':re.compile('(^ad-|^ads-)')})
+
+    [comment.extract() for comment in comments]
+    [tag.decompose() for tag in decomposable_tags]
+    [ad.decompose() for ad in ads]
```

Comment:
This isn't in-place. You can use
`comments = [comment.extract() for comment in comments]`